### PR TITLE
SE: Learn ObjectConstraint from IsNull operation

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Contributes to #5973

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/IsNull.cs [hunk 1]
```diff
@@ -30,5 +30,11 @@ internal static class IsNull
             && value.HasConstraint<ObjectConstraint>()
                 ? context.SetOperationConstraint(BoolConstraint.From(value.HasConstraint(ObjectConstraint.Null)))
                 : context.State;
+
+        public static ProgramState LearnBranchingConstraint(ProgramState state, IIsNullOperationWrapper isNull, bool useOpposite) =>
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/IsNull.cs:34]: I find "useOpposite" a bit confusing here. In this case, I would call it `notNullBranch` or something like that. In the SyntaxTree it is called "WhenNotNull".

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs [hunk 2]
```diff
@@ -43,6 +43,20 @@ public void FlasePositive(object o)
                 var b = o.ToString(); // Compliant, o is checked for null in this branch
             }
             catch (ApplicationException) when (o == null)
+            {
+                var b = o.ToString(); // Unreachable, o? cannot throw ApplicationException
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs:47]: But "ToString" could throw.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs [hunk 3]
```diff
@@ -62,109 +76,111 @@ public void TryCatch4(object o)
...
-    public class A
-    {
-      public bool booleanVal { get; set; }
-    }
+        public void Compliant(List<int> list)
+        {
+            var row = list?.Count;
+            if (row != null)
+            {
+                var type = list.ToArray();  // Noncompliant FP, nullability is inferred from result relation
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs:84]: Is this a TODO on our list? Consider adding a FIXME

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs [hunk 4]
```diff
@@ -62,109 +76,111 @@ public void TryCatch4(object o)
...
-      {
-        var type = list.ToArray(); // FIXME Non-compliant
-      }
-    }
+        public void Compliant1(List<int> list, A a)
+        {
+            var row = list?.Count;
+            if (a.booleanVal = (row != null))
+            {
+                var type = list.ToArray();  // Noncompliant FP, nullability is inferred from result relation
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs:100]: Same here

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs [hunk 5]
```diff
@@ -62,109 +76,111 @@ public void TryCatch4(object o)
...
 
-  public class ReproForIssue2338
-  {
-    public ConsoleColor Color { get; set; }
+        void Compliant2(object o)
+        {
+            switch (o?.GetHashCode())
+            {
+                case 1:
+                    o.ToString(); // Noncompliant FP, nullability is inferred from result relation
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs:127]: And here

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs [hunk 6]
```diff
@@ -62,109 +76,111 @@ public void TryCatch4(object o)
...
+            switch (obj?.Color)
+            {
+                case null:
+                    Console.ForegroundColor = obj.Color; // Noncompliant
+                    break;
+                case ConsoleColor.Red:
+                    Console.ForegroundColor = obj.Color; // Noncompliant FP, requires nullability support in captured flow operation
+                    break;
+                default:
+                    Console.WriteLine($"Color {obj.Color} is not supported."); // Noncompliant FP, requires nullability support in captured flow operation
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs:163]: This one is tricky.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs [hunk 7]
```diff
@@ -62,109 +76,111 @@ public void TryCatch4(object o)
...
+                    break;
+            }
+        }
+
+        public void Method2(ReproForIssue2338 obj)
+        {
+            obj = null;
+            switch (obj?.Color)
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs:171]: We need to learn that `obj?.Color` is always Null. Is there a TODO for this?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp8.cs [hunk 8]
```diff
@@ -52,7 +52,7 @@ public class NullablePrimitiveType
         public void NullablePrimitiveType_FN()
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp8.cs:52]:
```suggestion
        public void NullablePrimitiveType()
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs [hunk 9]
```diff
@@ -1061,7 +1061,7 @@ static void Main(string[] args)
                     continue;
                 }
 
-                var _ = providerCourse.Items.Where(items => items == "item1"); // FIXME Non-compliant FP
+                var _ = providerCourse.Items.Where(items => items == "item1"); // Noncompliant FP
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs:1064]: Can you mark where the issue is raised? Or do `providerCourse.Items.ToList()` to get rid of the confusing lambda.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Some tests could have more annotations.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/IsNull.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/IsNull.cs:34]: It's not perfect, but it follows the `useOpposite` from top level. So it indicates the fact that it's the flip/flop in negation chain.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs:47]: `ToString` is only invoked if `o` is `NotNull`, and the `catch ... when (o == null)` makes it unreachable.

This is actually a pretty advanced case.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs:84]: It leads to a topic called relationships. Should be addressed in [MMF-2563](https://jira.sonarsource.com/browse/MMF-2563) and I consider it fine for now. It's pretty advanced scenario

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs (hunk 6 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs:163]: Oh yes. I took some time to think about it. If we can already support it. But no, it explodes too much.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs (hunk 7 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp6.cs:171]: `obj?.Color` is nullable => MMF-2401

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs (hunk 9 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs:1064]: I'll do both. One to minimize the case. Other to be sure we don't do anything from the `Items?` stuff on line 1058

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM. (and merged)

**Pull request merged**